File: JS/twoNumberSum.js
array = [3, 5, -4, 8, 11, 1, -1, 6];
targetSum = 10;

array2 = [-21, 301, 12, 4, 65, 56, 210, 356, 9, -47];
targetSum2 = 164;

array3 = [];
targetSum3 = 0;

//O(n^2) time | O(1) space
// function twoNumberSum(array, targetSum) {
//   var hashN = new Map([]);
//   var length = array.length;
//   // Write your code here.
//   if (length > 0) {
//     //Loop through array, grab one number (x)
//     for (let i = 0; i < length; i++){
//       //Loop second time through array, grab another number (y) add to (x)
//       for (let j = 0; j < length; j++){
//         if (array[i] !== array[j]){
//           if (array[i] + array[j] === targetSum){
//             hashN.set(array[i], array[j]);
//             return hashN;
//           }
//         }
//       }
//     }
//   }
// }

//O(n) time | O(n) space
// function twoNumberSum(array, targetSum) {
//   var hashN = new Map([]);
//   var length = array.length;
//   // Write your code here.
//   if (length > 0) {
//     //Loop through array, grab one number (x)
//     for (let i = 0; i < length; i++){
//       //Loop second time through array, grab another number (y) add to (x)
//       for (let j = 0; j < length; j++){
//         if (array[i] !== array[j]){
//           if (targetSum - array[j] === array[i]){
//             hashN.set(array[i], array[j]);
//             return hashN;
//           }
//         }
//       }
//     }
//   }
// }

//Same As above with HashTable 
// function twoNumberSum(array, targetSum) {
//   const inArray = new Set();

//   for(let firstNum of array){
//     const secondNum = targetSum - firstNum;
//     if (inArray.has(secondNum)) {
//       return [firstNum, secondNum]
//     }
//     inArray.add(firstNum);
//   }
// }
  //Edge case: find other possibilities


//First sort
//Traverse both sides of array
//While left is less than right
//Add left to right = Sum
//Does Sum === targetSum
//O(nlog(n) | O(1))
function twoNumberSum(array, targetSum) {
  //Help Sort Function(compare a,b)
  var array = array.sort(function(a, b){return a - b});
  console.log(array);
  var left = 0;
  var right = array.length - 1;

  while (left < right){
    currentSum = array[left] + array[right]
    if (currentSum === targetSum){
      return [array[left], array[right]];
    } else if (currentSum < targetSum){
      left++;
    } else {
      right--;
    }
  }
  return [];
}

console.log(twoNumberSum(array, targetSum));
console.log(twoNumberSum(array2, targetSum2));
console.log(twoNumberSum(array3, targetSum3));

File: Python/twoNumberSum.py
array0 = [3,5,-4,8,11,1,-1,6]
targetSum0 = 10

array1 = [4,6]
targetSum1 = 10

array2 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 15]
targetSum2 = 18

array3 = [-21, 301, 12, 4, 65, 56, 210, 356, 9, -47]
targetSum3 = 164

array4 = [15]
targetSum4 = 15

# def twoNumberSum(array, targetSum):
#   for i in range(len(array)-1):
#     firstNum = array[i]
#     for j in range(i + 1, len(array)):
#       secondNum = array[j]
#       if firstNum + secondNum ==targetSum:
#         return [firstNum, secondNum]
#   return []

def twoNumberSum(array, targetSum):
    # Sort Array
    array.sort()
    # Set left index = 0 
    left = 0
    # Set right index = 0
    right = len(array) - 1

    # while right is greater than left
    while left < right:
        # currentSum = left element + right element
        currentSum = array[left] + array[right]
        # if currentSum = targetSum
        if currentSum == targetSum:
            # return left element, right element
            return [array[left], array[right]]
        # else if currentSum is less than targetSum, move to next left index 
        elif currentSum < targetSum:
            left += 1
        # else if currentSum is greater than targetSum, move to previous right index
        elif currentSum > targetSum:
            right -= 1
    # return array of left element and right element adding up to targetSum or return array with no elements
    return []



print(twoNumberSum(array0, targetSum0))
print(twoNumberSum(array1, targetSum1))
print(twoNumberSum(array2, targetSum2))
print(twoNumberSum(array3, targetSum3))
print(twoNumberSum(array4, targetSum4))